vendor/github.com/thanos-io/thanos/pkg/exthttp/hedging.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exthttp
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/caio/go-tdigest"
+	"github.com/cristalhq/hedgedhttp"
+)
+
+type CustomBucketConfig struct {
+	HedgingConfig HedgingConfig `yaml:"hedging_config"`
+}
+
+type HedgingConfig struct {
+	Enabled  bool    `yaml:"enabled"`
+	UpTo     uint    `yaml:"up_to"`
+	Quantile float64 `yaml:"quantile"`
+}
+
+func DefaultCustomBucketConfig() CustomBucketConfig {
+	return CustomBucketConfig{
+		HedgingConfig: HedgingConfig{
+			Enabled:  false,
+			UpTo:     3,
+			Quantile: 0.9,
+		},
+	}
+}
+
+type hedgingRoundTripper struct {
+	Transport http.RoundTripper
+	TDigest   *tdigest.TDigest
+	mu        sync.RWMutex
+	config    HedgingConfig
+}
+
+func (hrt *hedgingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+	resp, err := hrt.Transport.RoundTrip(req)
+	if err != nil {
+		return nil, err
+	}
+	duration := float64(time.Since(start).Milliseconds())
+	hrt.mu.Lock()
+	err = hrt.TDigest.Add(duration)
+	if err != nil {
+		return nil, err
```

Comment:
Potential deadlock here

Reply:
Thanks, I made a PR here https://github.com/thanos-io/thanos/pull/7962.
cc. @yeya24

Reply:
Thanks for the catch!  @damnever